 Naïve Bayes  

Praktikum Week 4 Sesi 3

1. Import Library


Kode Program

In [4]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris

# Load dataset Iris dari scikit-Learn
iris = load_iris()




Penjelasan

Pada tahap pertama dilakukan import library yang diperlukan untuk proses praktikum. Library numpy digunakan untuk melakukan pengolahan data numerik dan array, sedangkan pandas digunakan untuk mengolah data dalam bentuk DataFrame. Selanjutnya, load_iris dari Scikit-Learn digunakan untuk mengambil dataset Iris yang sudah tersedia di dalam library tersebut. Ketiga library ini menjadi dasar untuk melakukan persiapan dan pengolahan dataset sebelum model Naïve Bayes dibuat.                                                                                                                                                                                                                                                                                                                                       

Pada tahap ini dataset Iris dimuat menggunakan fungsi load_iris(). Dataset Iris merupakan dataset yang berisi karakteristik bunga berdasarkan empat fitur numerik, yaitu panjang sepal, lebar sepal, panjang petal, dan lebar petal. Dataset tersebut juga memiliki target berupa tiga spesies bunga, yaitu setosa, versicolor, dan virginica. Dataset Iris digunakan dalam praktikum karena fitur-fiturnya berupa data numerik kontinu sehingga sesuai untuk penerapan Gaussian Naïve Bayes.

Kode Program Mengubah Dataset Menjadi DataFrame

In [13]:
# Konversi data menjadi Dataframe pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]

# Definisikan X (fitur) dan y (target/label)
x = df.drop(columns=["Species"])
y = df["Species"]

print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Penjelasan

Dataset Iris yang telah dimuat kemudian dikonversi menjadi DataFrame menggunakan Pandas. iris.data berisi nilai dari empat fitur bunga, sedangkan iris.feature_names digunakan sebagai nama kolom. Setelah itu dibuat kolom tambahan bernama Species yang berisi nama spesies berdasarkan target dari dataset Iris. Dengan demikian, data menjadi lebih mudah dibaca dan dianalisis karena setiap fitur memiliki nama kolom yang jelas dan target klasifikasi ditampilkan dalam bentuk nama spesies.   

Pada tahap ini dataset dipisahkan menjadi dua bagian, yaitu variabel X sebagai fitur dan y sebagai target atau label. Variabel X diperoleh dengan menghapus kolom Species, sehingga berisi empat fitur numerik, yaitu sepal length, sepal width, petal length, dan petal width. Sementara itu, variabel y berisi nama spesies bunga yang akan diprediksi oleh model. Hasil yang ditampilkan menunjukkan bahwa dataset memiliki 150 baris data dan 4 fitur. Target y berisi kelas spesies Iris yang menjadi label klasifikasi.

Kode Program Melakukan Encoding pada Target

In [14]:
from sklearn.preprocessing import LabelEncoder

lab = LabelEncoder()
y = lab.fit_transform(y)

print(y)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


Penjelasan 

Karena target y masih berbentuk teks berupa nama spesies, dilakukan proses encoding menggunakan LabelEncoder. Encoding digunakan untuk mengubah label kategori menjadi bentuk numerik agar dapat digunakan dalam proses pelatihan model machine learning. Setelah proses encoding, setiap spesies memiliki representasi angka. Berdasarkan hasil pada modul, data setosa direpresentasikan sebagai 0, versicolor sebagai 1, dan virginica sebagai 2. Dengan demikian, target yang sebelumnya berupa teks sudah berubah menjadi data numerik yang dapat diproses oleh model Gaussian Naïve Bayes.

Kode Program Membagi Data Training dan Testing

In [15]:
from sklearn.model_selection import train_test_split

X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

Penjelasan 

Setelah fitur dan target dipersiapkan, dataset dibagi menjadi data training dan data testing menggunakan train_test_split. Parameter test_size=0.2 berarti 20% data digunakan sebagai data testing, sedangkan 80% sisanya digunakan sebagai data training. Data training digunakan untuk membangun dan melatih model, sedangkan data testing digunakan untuk mengetahui kemampuan model dalam melakukan prediksi terhadap data yang tidak digunakan selama proses pelatihan. Parameter random_state=0 digunakan agar pembagian data tetap konsisten ketika program dijalankan kembali.

Dataset Iris berisi fitur-fitur numerik kontinu seperti: 
Sepal Length (Panjang kelopak) 
Sepal Width (Lebar kelopak)
 Petal Length (Panjang mahkota)
 Petal Width (Lebar mahkota)

Kode Program Menentukan Jenis Naïve Bayes

In [16]:
from sklearn.naive_bayes import GaussianNB

model = GaussianNB()
NB_model = model.fit(X_train, y_train)

Penjelasan 

Sebelum membuat model, modul menjelaskan bahwa terdapat tiga jenis utama Naïve Bayes, yaitu Gaussian Naïve Bayes, Multinomial Naïve Bayes, dan Bernoulli Naïve Bayes. Gaussian Naïve Bayes digunakan untuk fitur numerik kontinu yang diasumsikan mengikuti distribusi Gaussian atau normal. Multinomial Naïve Bayes digunakan untuk data yang berbasis frekuensi kejadian, sedangkan Bernoulli Naïve Bayes digunakan untuk data biner atau data dengan nilai 0 dan 1. Karena seluruh fitur pada dataset Iris berupa data numerik kontinu, modul memilih Gaussian Naïve Bayes untuk proses klasifikasi.

Kode Program Membuat Model Gaussian Naïve Bayes

In [17]:
result = NB_model.predict(X_test)

result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

Penjelasan 

Pada tahap ini dibuat model menggunakan GaussianNB dari Scikit-Learn. GaussianNB merupakan implementasi Gaussian Naïve Bayes yang sesuai digunakan untuk dataset Iris karena seluruh fiturnya berupa data numerik kontinu. Objek model disimpan dalam variabel model, kemudian model dilatih menggunakan X_train dan y_train melalui fungsi fit(). Setelah proses tersebut selesai, model yang telah dilatih disimpan dalam variabel NB_model. Model inilah yang selanjutnya digunakan untuk melakukan prediksi terhadap data testing maupun data baru.

Kode Program Melakukan Prediksi pada Data Testing

In [21]:
from sklearn.metrics import accuracy_score

print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


Penjelasan 

Setelah model Gaussian Naïve Bayes selesai dilatih, dilakukan prediksi terhadap data testing menggunakan fungsi predict(). Data yang digunakan sebagai input adalah X_test, yaitu bagian data yang sebelumnya dipisahkan untuk pengujian. Hasil prediksi disimpan dalam variabel result. Berdasarkan modul, hasil prediksi berupa sekumpulan angka 0, 1, dan 2, yang masing-masing merepresentasikan kelas setosa, versicolor, dan virginica. Hasil prediksi ini nantinya dibandingkan dengan y_test untuk mengetahui tingkat akurasi model.


Pada tahap ini dilakukan evaluasi terhadap hasil prediksi menggunakan accuracy_score. Fungsi tersebut membandingkan label sebenarnya pada y_test dengan hasil prediksi yang tersimpan pada result. Hasil akurasi yang diperoleh adalah sekitar 96,67%. Artinya, pada data testing dan pembagian data yang digunakan dalam praktikum, model Gaussian Naïve Bayes mampu mengklasifikasikan sebagian besar data dengan benar. Nilai tersebut menunjukkan bahwa model memberikan hasil klasifikasi yang cukup baik pada dataset Iris dalam percobaan yang dilakukan pada modul.

Kode Program Prediksi Data Baru

In [22]:
input_data = [[5.1, 3.5, 1.4, 0.2]]

result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)

print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


d:\Perkuliahan\APK\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Penjelasan 

Pada tahap ini diberikan satu data baru dengan empat nilai fitur, yaitu [5.1, 3.5, 1.4, 0.2]. Data tersebut kemudian dimasukkan ke dalam model Gaussian Naïve Bayes menggunakan fungsi predict(). Hasil prediksi yang diperoleh masih berupa label numerik sehingga digunakan lab.inverse_transform() untuk mengembalikannya menjadi nama spesies. Berdasarkan hasil pada modul, data baru tersebut diprediksi sebagai setosa. Dengan demikian, model mengklasifikasikan karakteristik bunga dengan panjang sepal 5.1, lebar sepal 3.5, panjang petal 1.4, dan lebar petal 0.2 sebagai spesies setosa.


Warning tersebut muncul karena model Gaussian Naïve Bayes dilatih menggunakan X_train yang berasal dari DataFrame, sehingga model mengenali nama-nama fitur seperti sepal length (cm), sepal width (cm), petal length (cm), dan petal width (cm). Sementara itu, ketika melakukan prediksi menggunakan list atau NumPy array secara langsung, data input tidak mempunyai nama kolom. Oleh karena itu, Scikit-Learn memberikan peringatan bahwa data input tidak memiliki feature names yang sesuai dengan data ketika model dilatih. Modul menjelaskan bahwa solusi untuk menjaga struktur input tetap konsisten adalah mengubah data baru menjadi DataFrame dan memberikan nama kolom yang sama dengan fitur pada dataset Iris.

Kode Program Mengubah Input Menjadi DataFrame

In [24]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np. array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)

input_data = pd.DataFrame(
    np.array([[5.1, 3.5, 1.4, 0.2]]),
    columns=iris.feature_names
)

result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)

print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


Penjelasan 

Pada tahap ini data input yang sebelumnya berupa list diubah menjadi DataFrame menggunakan Pandas. Nama kolom diberikan menggunakan iris.feature_names, sehingga struktur data baru menjadi sama dengan struktur fitur yang digunakan ketika model Gaussian Naïve Bayes dilatih. Setelah itu data diprediksi menggunakan NB_model.predict(), kemudian hasil numeriknya dikembalikan menjadi nama spesies menggunakan lab.inverse_transform(). Hasil prediksi tetap menunjukkan bahwa data [5.1, 3.5, 1.4, 0.2] termasuk ke dalam spesies setosa. Perbedaannya adalah pada tahap ini input memiliki nama fitur yang sesuai sehingga warning mengenai feature names dapat dihindari.

Tugas   

                                                                                                                                                                                                                                                                                                                                      1. Ikuti semua tahapan dalam modul ini.    

                                                                                                                                                             2. Prediksi data baru masuk ke species apa: [7.3, 3.8, 4.6, 2.3]

Tugas 2 — Prediksi Species Iris

Kode Program

In [27]:
input_data = pd.DataFrame(
    np.array([[7.3, 3.8, 4.6, 2.3]]),
    columns=iris.feature_names
)

result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)

print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['virginica']


Penjelasan 

Pada tugas ini dilakukan prediksi terhadap satu data baru pada dataset Iris menggunakan model Gaussian Naïve Bayes yang telah dibangun pada bagian praktikum. Data yang digunakan adalah [7.3, 3.8, 4.6, 2.3], yang terdiri dari empat nilai sesuai dengan empat fitur pada dataset Iris, yaitu sepal length, sepal width, petal length, dan petal width. Karena model sebelumnya telah dilatih menggunakan empat fitur tersebut, data baru dimasukkan ke dalam DataFrame dan diberikan nama kolom yang sama dengan iris.feature_names. Cara ini juga menjaga agar struktur data input sesuai dengan struktur data yang digunakan ketika model dilatih.

Selanjutnya, fungsi NB_model.predict(input_data) digunakan untuk menentukan kelas dari data baru berdasarkan pola yang telah dipelajari oleh model Gaussian Naïve Bayes. Hasil prediksi awal berupa label numerik, sehingga digunakan lab.inverse_transform(result) untuk mengubah label tersebut kembali menjadi nama spesies Iris. Berdasarkan hasil prediksi, data [7.3, 3.8, 4.6, 2.3] diklasifikasikan sebagai versicolor.